configs/components/_base-module.rb

```diff
@@ -0,0 +1,71 @@
+# This file is a basis for vendor modules that should be installed alongside
+# puppet-agent in settings[:module_vendordir]. It should not be used as a
+# component; Instead, other module components should load it using
+# `instance_eval`. Here's an example:
+#
+# component "module-puppetlabs-stdlib" do |pkg, settings, platform|
+#   pkg.version "4.25.1"
+#
+#   instance_eval File.read("configs/components/_base-module.rb")
+# end
+#
+# The module's author and name can be inferred from the name of the component
+# when it has the format "module-<author>-<name>", but you may set them
+# explicitly instead by defining `module_author` and `module_name` before
+# loading this file.
+#
+# No dependency resolution is attempted automatically; You must define separate
+# vanagon components and make their dependencies explicit.
+
+unless defined?(pkg)
+  raise "_base-module.rb is not a valid vanagon component; Load it with `instance_eval` in another component instead."
+end
+
+unless defined?(module_author) && defined?(module_name)
+  # If the module_author and module_name aren't already defined in the component
+  # and the component's name matches the format "module-<author>-<name>",
+  # assume we can extract the author and name that way
+  if match_data = pkg.get_name.match(/module-([^-]+)-([^-]+)/)
+    module_author, module_name = match_data.captures
+  else
+    raise "You must supply the module name and author for #{pkg.get_name}"
+  end
+end
+
+unless pkg.get_version
+  raise "You must set a version for the #{module_author}-#{module_name} module."
+end
+
+module_slug ||= "#{module_author}-#{module_name}-#{pkg.get_version}"
+
+# Fetch module metadata from the forge
+require 'net/http'
```

Comment:
Not sure requesting metadata from the forge is ideal -- the tarball URLs are consistently named, so we could also construct them without making any API requests instead of doing it this way. We'd need to manually supply the md5sum in that case, though.